src/bucket/BucketSnapshot.cpp

```diff
@@ -125,26 +138,40 @@ BucketSnapshot::loadKeysWithLimits(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
         {
             auto [entryOp, bloomMiss] = getEntryAtOffset(
                 *currKeyIt, *offOp, mBucket->getIndex().getPageSize());
+
             if (entryOp)
             {
-                if (entryOp->type() != DEADENTRY)
+                // Don't return tombstone entries, as these do not exist wrt
+                // ledger state
+                if (!BucketT::isTombstoneEntry(*entryOp))
                 {
-                    bool addEntry = true;
-                    if (lkMeter)
+                    // Only live bucket loads can be metered
+                    if constexpr (std::is_same_v<BucketT, LiveBucket>)
                     {
-                        // Here, we are metering after the entry has been
-                        // loaded. This is because we need to know the size of
-                        // the entry to meter it. Future work will add metering
-                        // at the xdr level.
-                        auto entrySize = xdr::xdr_size(entryOp->liveEntry());
-                        addEntry = lkMeter->canLoad(*currKeyIt, entrySize);
-                        lkMeter->updateReadQuotasForKey(*currKeyIt, entrySize);
+                        bool addEntry = true;
+                        if (lkMeter)
+                        {
+                            // Here, we are metering after the entry has been
+                            // loaded. This is because we need to know the size
+                            // of the entry to meter it. Future work will add
+                            // metering at the xdr level.
+                            auto entrySize =
+                                xdr::xdr_size(entryOp->liveEntry());
+                            addEntry = lkMeter->canLoad(*currKeyIt, entrySize);
+                            lkMeter->updateReadQuotasForKey(*currKeyIt,
+                                                            entrySize);
+                        }
+                        if (addEntry)
+                        {
+                            result.push_back(entryOp->liveEntry());
+                        }
                     }
-                    if (addEntry)
+                    else
                     {
-                        result.push_back(entryOp->liveEntry());
+                        result.push_back(*entryOp);
```

Comment:
I'm not sure I understand this, why do we add entry unconditionally as it seems like it must have been archived at this point? Is that for `getLedgerEntry` endpoint?

Reply:
Archival is not enforced at the database level, but at the txn level. Core needs to query about Archival State for a variety of reasons. For example, to check if a new entry creation is valid, core needs to check that an archived version of that key does not exist in the Live bucketlist or the archival bucketlists, so we can't just return null whenever an entry is archived at the db level.